bson/bsoncodec/default_value_decoders.go

```diff
@@ -1504,14 +1504,14 @@ func (dvd DefaultValueDecoders) UnmarshalerDecodeValue(dc DecodeContext, vr bson
 		return err
 	}
 
-	// If the target Go value is a pointer and the BSON field value is empty, set the value to the
+	// If the target Go value is an addressable pointer and the BSON field value is empty, set the value to the
 	// zero value of the pointer (nil) and don't call UnmarshalBSON. UnmarshalBSON has no way to
 	// change the pointer value from within the function (only the value at the pointer address),
 	// so it can't set the pointer to "nil" itself. Since the most common Go value for an empty BSON
 	// field value is "nil", we set "nil" here and don't call UnmarshalBSON. This behavior matches
 	// the behavior of the Go "encoding/json" unmarshaler when the target Go value is a pointer and
 	// the JSON field value is "null".
-	if val.Kind() == reflect.Ptr && len(src) == 0 {
+	if val.CanSet() && val.Kind() == reflect.Ptr && len(src) == 0 {
```

Comment:
While this results in the correct behavior, it's really difficult to understand _why_ it works. It seems like `val.CanSet()` will basically always be false if the `if !val.Type().Implements(tUnmarshaler)` block above is run because that block converts the value to a pointer (`val = val.Addr()`). Getting the `Addr()` allows us to call an `UnmarshalBSON` function with a pointer receiver, but it also makes the value unsettable (see an example [here](https://go.dev/play/p/PFN3f5PCk1b)). If we ran the `if !val.Type().Implements(tUnmarshaler)` block above, we should never run this block.

I think a more obvious solution is to move this block above the `if !val.Type().Implements(tUnmarshaler)` block ([line 1488](https://github.com/mongodb/mongo-go-driver/blob/62908c2132e1a88816eebd8cc79ed56fd1e2c5c6/bson/bsoncodec/default_value_decoders.go#L1462)) so we can check if `val` is a pointer _before_ it may be converted to a pointer.

E.g.
```go
if !val.IsValid() || (!val.Type().Implements(tUnmarshaler) && !reflect.PtrTo(val.Type()).Implements(tUnmarshaler)) {
	return ValueDecoderError{Name: "UnmarshalerDecodeValue", Types: []reflect.Type{tUnmarshaler}, Received: val}
}

_, src, err := bsonrw.Copier{}.CopyValueToBytes(vr)
if err != nil {
	return err
}

// If the target Go value ...
if val.Kind() == reflect.Ptr && len(src) == 0 {
	if val.CanSet() {
		val.Set(reflect.Zero(val.Type()))
	}
	return nil
}
// ...
```

The important difference is the `val.Kind() == reflect.Ptr` is intended to check if the _struct field_ is a pointer type. When `val` is explicitly set to be a pointer (`val = val.Addr()`), it changes the meaning of the pointer check (i.e. it's not checking if the struct field is a pointer anymore). If the struct field is not a pointer type, then the "skip UnmarshalBSON for null values" logic is no longer relevant.

Reply:
Great suggestion, thank you.